platforms/compare_test.go

```diff
@@ -0,0 +1,205 @@
+				true: {
+					"linux/386",
+				},
+				false: {
+					"linux/amd64",
+					"linux/arm/v7",
+					"linux/arm64",
+					"windows/amd64",
+					"windows/arm",
+				},
+			},
+		},
+		{
+			platform: "windows/amd64",
+			matches: map[bool][]string{
+				true: {"windows/amd64"},
+				false: {
+					"linux/amd64",
+					"linux/arm/v7",
+					"linux/arm64",
+					"windows/arm",
+				},
+			},
+		},
+		{
+			platform: "linux/arm/v8",
+			matches: map[bool][]string{
+				true: {
+					"linux/arm",
+					"linux/arm/v5",
+					"linux/arm/v6",
+					"linux/arm/v7",
+					"linux/arm/v8",
+				},
+				false: {
+					"linux/amd64",
+					"linux/arm/v4",
+					"linux/arm64",
+					"windows/amd64",
+					"windows/arm",
+				},
+			},
+		},
+		{
+			platform: "linux/arm/v7",
+			matches: map[bool][]string{
+				true: {
+					"linux/arm",
+					"linux/arm/v5",
+					"linux/arm/v6",
+					"linux/arm/v7",
+				},
+				false: {
+					"linux/amd64",
+					"linux/arm/v4",
+					"linux/arm/v8",
+					"linux/arm64",
+					"windows/amd64",
+					"windows/arm",
+				},
+			},
+		},
+		{
+			platform: "linux/arm/v6",
+			matches: map[bool][]string{
+				true: {
+					"linux/arm/v5",
+					"linux/arm/v6",
+				},
+				false: {
+					"linux/amd64",
+					"linux/arm",
+					"linux/arm/v4",
+					"linux/arm/v7",
+					"linux/arm/v8",
+					"linux/arm64",
+					"windows/amd64",
+					"windows/arm",
+				},
+			},
+		},
+		{
+			platform: "linux/arm/v5",
+			matches: map[bool][]string{
+				true: {
+					"linux/arm/v5",
+				},
+				false: {
+					"linux/amd64",
+					"linux/arm",
+					"linux/arm/v4",
+					"linux/arm/v6",
+					"linux/arm/v7",
+					"linux/arm/v8",
+					"linux/arm64",
+					"windows/amd64",
+					"windows/arm",
+				},
+			},
+		},
+		{
+			platform: "linux/arm/v4",
+			matches: map[bool][]string{
+				true: {
+					"linux/arm/v4",
+				},
+				false: {
+					"linux/amd64",
+					"linux/arm",
+					"linux/arm/v5",
+					"linux/arm/v6",
+					"linux/arm/v7",
+					"linux/arm/v8",
+					"linux/arm64",
+					"windows/amd64",
+					"windows/arm",
+				},
+			},
+		},
+		{
+			platform: "linux/arm64",
+			matches: map[bool][]string{
+				true: {
+					"linux/arm64",
+					"linux/arm64/v8",
+				},
+				false: {
+					"linux/amd64",
+					"linux/arm",
+					"linux/arm/v4",
+					"linux/arm/v5",
+					"linux/arm/v6",
+					"linux/arm/v7",
+					"linux/arm/v8",
+					"linux/arm/v9",
+					"linux/arm64/v9",
+					"windows/amd64",
+					"windows/arm",
+				},
+			},
+		},
+	} {
+		p, err := Parse(testcase.platform)
+		if err != nil {
+			t.Fatal(err)
+		}
+		m := Only(p)
+		for shouldMatch, platforms := range testcase.matches {
+			for _, matchPlatform := range platforms {
+				mp, err := Parse(matchPlatform)
+				if err != nil {
+					t.Fatal(err)
+				}
+				if match := m.Match(mp); shouldMatch != match {
+					t.Errorf("Only(%q).Match(%q) should return %v, but returns %v", testcase.platform, matchPlatform, shouldMatch, match)
+				}
+			}
+		}
```

Comment:
Perhaps would be nice to use subtests?

```suggestion
		tc := testcase
		t.Run(tc.platform, func(t *testing.T) {
			p, err := Parse(testcase.platform)
			if err != nil {
				t.Fatal(err)
			}
			m := Only(p)
			for shouldMatch, platforms := range tc.matches {
				for _, matchPlatform := range platforms {
					mp, err := Parse(matchPlatform)
					if err != nil {
						t.Fatal(err)
					}
					if match := m.Match(mp); shouldMatch != match {
						t.Errorf("Only(%q).Match(%q) should return %v, but returns %v", tc.platform, matchPlatform, shouldMatch, match)
					}
				}
			}
		})
```

```console
=== RUN   TestOnly
--- PASS: TestOnly (0.00s)
=== RUN   TestOnly/linux/amd64
    --- PASS: TestOnly/linux/amd64 (0.00s)
=== RUN   TestOnly/linux/386
    --- PASS: TestOnly/linux/386 (0.00s)
=== RUN   TestOnly/windows/amd64
    --- PASS: TestOnly/windows/amd64 (0.00s)
=== RUN   TestOnly/linux/arm/v8
    --- PASS: TestOnly/linux/arm/v8 (0.00s)
=== RUN   TestOnly/linux/arm/v7
    --- PASS: TestOnly/linux/arm/v7 (0.00s)
=== RUN   TestOnly/linux/arm/v6
    --- PASS: TestOnly/linux/arm/v6 (0.00s)
=== RUN   TestOnly/linux/arm/v5
    --- PASS: TestOnly/linux/arm/v5 (0.00s)
=== RUN   TestOnly/linux/arm/v4
    --- PASS: TestOnly/linux/arm/v4 (0.00s)
=== RUN   TestOnly/linux/arm64
    --- PASS: TestOnly/linux/arm64 (0.00s)
PASS

Process finished with exit code 0
```

Reply:
Oooh, I like that.
(Never used subtests before! :+1:)

Reply:
https://github.com/containerd/containerd/compare/7d8b50ea1d593c27bae757e29ffd58d19648f85e..66a3f2cbc77d2e0654f6faad01171643a28ce1c8?w=1 :eyes: